File: Shortest Source to Destination Path[2-August-2023]/solution.cpp
//  >> C++ CODE <<



class Solution {
  public:
    int shortestDistance(int N, int M, vector<vector<int>> A, int X, int Y) {
        
        if(A[0][0]==0) return -1;
        queue<pair<int,int>> q;
        vector<vector<bool>> vis(N,vector<bool>(M,false));
        int path=0;
        q.push({0,0});
        vis[0][0]=true;
        while(!q.empty()) {
            int size=q.size();
            for(int i=0;i<size;i++) {
                auto cur=q.front();
                q.pop();
                if(cur.first==X && cur.second==Y) return path;
                if(cur.first+1<N && !vis[cur.first+1][cur.second] && A[cur.first+1][cur.second]==1) {
                    q.push({cur.first+1,cur.second});
                    vis[cur.first+1][cur.second]=true;
                }
                if(cur.first-1>=0 && !vis[cur.first-1][cur.second] && A[cur.first-1][cur.second]==1) {
                    q.push({cur.first-1,cur.second});
                    vis[cur.first-1][cur.second]=true;
                }
                if(cur.second+1<M && !vis[cur.first][cur.second+1] && A[cur.first][cur.second+1]==1) {
                    q.push({cur.first,cur.second+1});
                    vis[cur.first][cur.second+1]=true;
                }
                if(cur.second-1>=0 && !vis[cur.first][cur.second-1] && A[cur.first][cur.second-1]==1) {
                    q.push({cur.first,cur.second-1});
                    vis[cur.first][cur.second-1]=true;
                }
            }
            path++;
        }
        return -1;
        
    }

};

File: Shortest Source to Destination Path[2-August-2023]/Solution.py
#  >> PYTHON CODE <<




from queue import Queue

class Solution:
    
    def shortestDistance(self,N,M,A,X,Y):
        
        if(A[0][0]==0): return -1
        
        queue=Queue()
        visited=[[False for i in range(M)] for j in range(N)]
        path=0;

        queue.put([0,0])
        visited[0][0]=True
        while(not queue.empty()):
            
            size=queue.qsize()
            
            for i in range(size):
                cur=queue.get()
                
                if(cur[0]==X and cur[1]==y): return path
                
                if(cur[0]+1<N and (not visited[cur[0]+1][cur[1]]) and A[cur[0]+1][cur[1]]==1):
                    queue.put([cur[0]+1,cur[1]])
                    visited[cur[0]+1][cur[1]]=True
                if(cur[0]-1>=0 and (not visited[cur[0]-1][cur[1]]) and A[cur[0]-1][cur[1]]==1):
                    queue.put([cur[0]-1,cur[1]])
                    visited[cur[0]-1][cur[1]]=True
                if(cur[1]+1<M and (not visited[cur[0]][cur[1]+1]) and A[cur[0]][cur[1]+1]==1):
                    queue.put([cur[0],cur[1]+1])
                    visited[cur[0]][cur[1]+1]=True
                if(cur[1]-1>=0 and (not visited[cur[0]][cur[1]-1]) and A[cur[0]][cur[1]-1]==1):
                    queue.put([cur[0],cur[1]-1])
                    visited[cur[0]][cur[1]-1]=True

            
            
            path+=1
            
            
        return -1